src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/CannotChangeParameterName.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    /// <summary>
+    /// This class implements a rule to check that the parameter names between public methods do not change.
+    /// </summary>
+    public class CannotChangeParameterName : IRule
+    {
+        private readonly RuleSettings _settings;
+
+        public CannotChangeParameterName(RuleSettings settings, IRuleRegistrationContext context, bool enableRuleCannotChangeParameterName)
+        {
+            _settings = settings;
+            if (enableRuleCannotChangeParameterName)
+            {
+                context.RegisterOnMemberSymbolAction(RunOnMemberSymbol);
+            }
+        }
+
+        private void RunOnMemberSymbol(
+            ISymbol? left,
+            ISymbol? right,
+            ITypeSymbol leftContainingType,
+            ITypeSymbol rightContainingType,
+            MetadataInformation leftMetadata,
+            MetadataInformation rightMetadata,
+            IList<CompatDifference> differences)
+        {
+            if (left is not IMethodSymbol leftMethod || right is not IMethodSymbol rightMethod)
+            {
+                return;
+            }
+
+            for (int i = 0; i < leftMethod.Parameters.Length; i++)
+            {
+                IParameterSymbol leftParam = leftMethod.Parameters[i];
+                IParameterSymbol rightParam = rightMethod.Parameters[i];
```

Comment:
Don't you need to check that the rightMethod.Parameters.Length is equal to leftMethod.Parameters.Length?

Reply:
I thought so too, but when testing, I couldn't get the RuleRunner to pass in methods with a different number of parameters, even when the methods were otherwise identical.

However, I can add a check defensively. It just seems that the RuleRunner is taking care of that issue for us by not invoking the RunOnMemberSymbol for methods with a different number of parameters.

Reply:
Yes, that's handled by the mapping system (more precisely by the passed in equality comparer): https://github.com/dotnet/sdk/blob/1b6da85abc592947ca0bb4c7de01d106466258da/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Abstractions/Mappers/TypeMapper.cs#L138.

That said, it's an "outside" component (in the meaning of outside the rule), therefor I would add a Debug.Assert check to formalize that requirement in debug code.